pennylane/qnodes/base.py

```diff
@@ -423,8 +474,20 @@ def qfunc(a, w):
         arg_vars = [Variable(idx, name) for idx, name in enumerate(_flatten(variable_name_strings))]
         self.num_variables = len(arg_vars)
 
-        # arrange the newly created Variables in the nested structure of args
-        arg_vars = unflatten(arg_vars, args)
+        # Arrange the newly created Variables in the nested structure of args.
+        # Make sure that NumPy scalars are converted into Python scalars.
+        arg_vars = [
+            i.item() if isinstance(i, np.ndarray) and i.ndim == 0 else i
+            for i in unflatten(arg_vars, args)
+        ]
+
+        if self._trainable_args is not None and len(self._trainable_args) != len(args):
+            # If some of the input arguments are marked as non-differentiable,
+            # then replace the variable instances in arg_vars back with the
+            # original objects.
+            for a, _ in enumerate(args):
+                if a not in self._trainable_args:
+                    arg_vars[a] = args[a]
 
         # kwargs
         # if not mutable: must convert auxiliary arguments to named Variables so they can be updated without re-constructing the circuit
```

Comment:
You should add some unit tests for these changes to the qnode

Reply:
Done!